File: Practica_1_1/Practica_1_1.m
%clear all;
close all;
clc;
%% Información de las canciones seleccionadas
% Come_As_You_Are.wav Fs_rock= 48000Hz a 16 bits
% Nocturne_in_C.wav Fs_clasic = 48000Hz a 16 bits
% WILDFLOWER.wav Fs_pop = 48000Hz a 16 bits
%% Para saber frecuencia y bits por muestra
% [y, Fs_Come] = audioread('Come_As_You_Are.wav');
% info = audioinfo('WILDFLOWER.wav');
% disp(info.BitsPerSample);
% 
% [y, Fs_Nocturne] = audioread('Nocturne_No_2.wav');
% info = audioinfo('WILDFLOWER.wav');
% disp(info.BitsPerSample);
% 
% [y, Fs_Wilflower] = audioread('WILDFLOWER.wav');
% info = audioinfo('WILDFLOWER.wav');
% disp(info.BitsPerSample);
%% Inicio de práctica 1.1

% Lista de archivos y sus géneros (puedes ajustarlos)
audioFiles = {'Come_As_You_Are.wav', 'Nocturne_in_C.wav', 'WILDFLOWER.wav'};
genres = {'Rock', 'Clásica', 'Pop'};

nBits = 16;              % Usamos 16 bits
nLevels = 2^nBits;       % 65,536 niveles

% Inicializamos un vector para guardar la entropía de cada archivo
entropies = zeros(length(audioFiles), 1);

for i = 1:length(audioFiles)
    filename = audioFiles{i};
    [x, fs] = audioread(filename);
    
    % Si el audio es estéreo, seleccionamos solo el primer canal
    if size(x, 2) > 1
        x = x(:, 1);
    end
    
    % Normalización: escalamos la señal para que el máximo valor absoluto sea 1
    x = x / max(abs(x));
    
    % Cuantización a 16 bits: mapeamos de [-1,1] a [0, 65535]
    x_quant = floor((x + 1) / 2 * (nLevels - 1));
    
    % Definir el alfabeto: niveles de 0 a 65535
    alfabeto = (0:nLevels-1)';  % vector columna
    
    % Contar ocurrencias de cada nivel
    counts = histcounts(x_quant, -0.5:1:(nLevels-1)+0.5);
    totalMuestras = length(x_quant);
    probabilities = counts / totalMuestras;
    
    % Calcular la entropía (solo se consideran los niveles que aparecen, es decir, P>0)
    nonzero = probabilities > 0;
    H = -sum(probabilities(nonzero) .* log2(probabilities(nonzero)));
    entropies(i) = H;
    
    fprintf('Archivo: %s (Género: %s)\n', filename, genres{i});
    fprintf('Frecuencia de muestreo: %d Hz, Total de muestras: %d\n', fs, totalMuestras);
    fprintf('Entropía: %.4f bits/símbolo\n\n', H);
    
    % Guardar la tabla completa de probabilidades
    T = table(alfabeto, probabilities', 'VariableNames', {'Nivel', 'Probabilidad'});
    csvFileName = strcat(filename(1:end-4), '_tabla_probabilidades.csv');
    writetable(T, csvFileName);
    
    % Extraer los 7 niveles con mayor probabilidad
    [sortedProbs, sortedIndices] = sort(probabilities, 'descend');
    top7Levels = alfabeto(sortedIndices(1:7));
    top7Probs = sortedProbs(1:7);
    % Calcular la información para esos niveles
    top7Info = log2(1 ./ top7Probs);
    
    % Extraer los 7 niveles con menor probabilidad (descartando los que no aparecen)
    nonzeroIndices = find(probabilities > 0); % índices de niveles presentes
    nonzeroProbs = probabilities(probabilities > 0);
    [sortedNonzeroProbs, sortedNonzeroOrder] = sort(nonzeroProbs, 'ascend');
    bottom7Levels = alfabeto(nonzeroIndices(sortedNonzeroOrder(1:7)));
    bottom7Probs = sortedNonzeroProbs(1:7);
    bottom7Info = log2(1 ./ bottom7Probs);
    
    % Mostrar en consola las tablas de top7 y bottom7
    T_top7 = table(top7Levels, top7Probs', top7Info', 'VariableNames', {'Nivel', 'Probabilidad', 'Informacion'});
    T_bottom7 = table(bottom7Levels, bottom7Probs', bottom7Info', 'VariableNames', {'Nivel', 'Probabilidad', 'Informacion'});
    disp(['Archivo: ', filename, ' - Top 7 niveles:']);
    disp(T_top7);
    disp(['Archivo: ', filename, ' - Bottom 7 niveles:']);
    disp(T_bottom7);
    
    % Graficar para este audio los 7 niveles con mayor y menor probabilidad
    figure;
    subplot(1,2,1)
    bar(top7Levels, top7Info, 'FaceColor', [0.2 0.6 0.5])
    xlabel('Nivel')
    ylabel('Información (bits)')
    title(sprintf('Top 7 niveles - %s', genres{i}))
    if(strcmp(filename, 'Come_As_You_Are.wav'))
        xlim([32760 32773])
    end
    grid on;
    
    subplot(1,2,2)
    bar(bottom7Levels, bottom7Info, 'FaceColor', [0.8 0.4 0.4])
    xlabel('Nivel')
    ylabel('Información (bits)')
    title(sprintf('Bottom 7 niveles - %s', genres{i}))
    grid on;
end

% Comparar las entropías de cada género en un gráfico de barras
figure;
bar(categorical(genres), entropies, 'FaceColor', [0.4 0.6 0.8])
xlabel('Género')
ylabel('Entropía (bits/símbolo)')
title('Comparación de Entropías entre Géneros')
grid on;

File: Practica_1_1/Practica1_1.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile
import pandas as pd
import os

# Configuración inicial
audio_files = ['Come_As_You_Are.wav', 'Nocturne_in_C.wav', 'WILDFLOWER.wav']
genres = ['Rock', 'Clásica', 'Pop']
n_bits = 16
n_levels = 2 ** n_bits  # 65536 niveles
entropies = []

for idx, filename in enumerate(audio_files):
    # Leer archivo de audio
    fs, data = wavfile.read(filename)
    
    # Convertir a mono si es estéreo
    if len(data.shape) > 1:
        data = data[:, 0]
        
    # Normalización
    data_norm = data / np.max(np.abs(data))
    
    # Cuantización a 16 bits
    data_quant = np.floor((data_norm + 1) / 2 * (n_levels - 1)).astype(int)
    
    # Crear alfabeto
    alfabeto = np.arange(n_levels)
    
    # Calcular histograma
    counts, _ = np.histogram(data_quant, bins=np.arange(-0.5, n_levels + 0.5))
    total_muestras = len(data_quant)
    probabilities = counts / total_muestras
    
    # Calcular entropía
    non_zero = probabilities > 0
    H = -np.sum(probabilities[non_zero] * np.log2(probabilities[non_zero]))
    entropies.append(H)
    
    # Imprimir información
    print(f'Archivo: {filename} (Género: {genres[idx]})')
    print(f'Frecuencia de muestreo: {fs} Hz, Total de muestras: {total_muestras}')
    print(f'Entropía: {H:.4f} bits/símbolo\n')
    
    # Guardar tabla de probabilidades
    df = pd.DataFrame({'Nivel': alfabeto, 'Probabilidad': probabilities})
    csv_filename = os.path.splitext(filename)[0] + '_tabla_probabilidades.csv'
    df.to_csv(csv_filename, index=False)
    
    # Niveles con mayor probabilidad
    sorted_indices = np.argsort(probabilities)[::-1]
    top7_levels = alfabeto[sorted_indices[:7]]
    top7_probs = probabilities[sorted_indices[:7]]
    top7_info = np.log2(1 / top7_probs)
    
    # Niveles con menor probabilidad (no cero)
    non_zero_indices = np.where(probabilities > 0)[0]
    non_zero_probs = probabilities[non_zero_indices]
    sorted_non_zero = np.argsort(non_zero_probs)
    bottom7_levels = non_zero_indices[sorted_non_zero[:7]]
    bottom7_probs = non_zero_probs[sorted_non_zero[:7]]
    bottom7_info = np.log2(1 / bottom7_probs)
    
    # Mostrar tablas
    print(f'Archivo: {filename} - Top 7 niveles:')
    print(pd.DataFrame({'Nivel': top7_levels, 
                       'Probabilidad': top7_probs, 
                       'Informacion': top7_info}))
    print(f'\nArchivo: {filename} - Bottom 7 niveles:')
    print(pd.DataFrame({'Nivel': bottom7_levels, 
                       'Probabilidad': bottom7_probs, 
                       'Informacion': bottom7_info}))
    
    # Graficar
    plt.figure(figsize=(12, 6))
    
    # Subplot izquierdo (Top 7)
    plt.subplot(1, 2, 1)
    plt.bar(top7_levels, top7_info, color='#33CC99')
    plt.xlabel('Nivel')
    plt.ylabel('Información (bits)')
    plt.title(f'Top 7 niveles - {genres[idx]}')
    if filename == 'Come_As_You_Are.wav':
        plt.xlim(32760, 32773)
    plt.grid(True)
    
    # Subplot derecho (Bottom 7)
    plt.subplot(1, 2, 2)
    plt.bar(bottom7_levels, bottom7_info, color='#FF6666')
    plt.xlabel('Nivel')
    plt.ylabel('Información (bits)')
    plt.title(f'Bottom 7 niveles - {genres[idx]}')
    plt.grid(True)
    
    plt.tight_layout()
    plt.show()

# Gráfico comparativo de entropías
plt.figure()
plt.bar(genres, entropies, color='#66B2FF')
plt.xlabel('Género')
plt.ylabel('Entropía (bits/símbolo)')
plt.title('Comparación de Entropías entre Géneros')
plt.grid(True)
plt.show()